substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/SubstitutionReportFeature.java

```diff
@@ -0,0 +1,188 @@
+        AnalysisUniverse universe = access.getUniverse();
+        for (AnalysisMethod method : universe.getMethods()) {
+            if (method.wrapped instanceof SubstitutionMethod) {
+                SubstitutionMethod subMethod = (SubstitutionMethod) method.wrapped;
+                if (subMethod.isUserSubstitution()) {
+                    String jarLocation = getJarLocation(subMethod.getAnnotated().getDeclaringClass());
+                    substitutions.putIfAbsent(jarLocation, new Substitutions());
+                    substitutions.get(jarLocation).addMethod(subMethod);
+                }
+            }
+        }
+    }
+
+    private void findSubstitutedFields(FeatureImpl.AfterAnalysisAccessImpl access) {
+        AnalysisUniverse universe = access.getUniverse();
+        for (AnalysisField field : universe.getFields()) {
+            if (field.wrapped instanceof SubstitutionField) {
+                SubstitutionField subField = (SubstitutionField) field.wrapped;
+                if (subField.isUserSubstitution()) {
+                    String jarLocation = getJarLocation(subField.getAnnotated().getDeclaringClass());
+                    substitutions.putIfAbsent(jarLocation, new Substitutions());
+                    substitutions.get(jarLocation).addField(subField);
+                }
+            }
+        }
+    }
+
+    private void reportSubstitutions() {
+        ReportUtils.report("substitutions performed by native-image","reports", "substitutions", "csv", pw -> {
+            pw.println("location , category (type/method/field) , original , annotated");
+            for (Map.Entry<String, Substitutions> g : substitutions.entrySet()) {
+                for (Map.Entry<ResolvedJavaType, ResolvedJavaType> e : g.getValue().getSubstitutedTypes().entrySet()) {
+                    pw.println(formatSubstitution(g.getKey(), "type", e.getKey(), e.getValue(), t -> t.toJavaName(true)));
+                }
+                for (Map.Entry<ResolvedJavaMethod, ResolvedJavaMethod> e : g.getValue().getSubstitutedMethods().entrySet()) {
+                    pw.println(formatSubstitution(g.getKey(), "method", e.getKey(), e.getValue(), this::formatMethod));
+                }
+                for (Map.Entry<ResolvedJavaField, ResolvedJavaField> e : g.getValue().getSubstitutedFields().entrySet()) {
+                    pw.println(formatSubstitution(g.getKey(), "field", e.getKey(), e.getValue(), this::formatField));
+                }
+            }
+        });
+    }
+
+    private String getJarLocation(ResolvedJavaType type) {
+        Class<?> annotatedClass = OriginalClassProvider.getJavaClass(GraalAccess.getOriginalSnippetReflection(), type);
+        CodeSource source = annotatedClass.getProtectionDomain().getCodeSource();
+        return source == null ? "unknown" : source.getLocation().toString();
+    }
+
+    private String formatMethod(ResolvedJavaMethod method) {
```

Comment:
ResolvedJavaMethod#format should also be a oneliner.

Reply:
If you are talking about `formatMethod` method, it already is?

Reply:
Sure, but `format` on `ResolvedJavaMethod` already exists.

Reply:
I missed that, fixed.